Golden Path documentation (#149)
Unity-Technologies/com.unity.multiplayer.docs: docs/tutorials/goldenpath.md

---
id: goldenpath

> title:  Hello World to Goldenpath

Author: Title updated to 

Building Golden Path

description: Tutorial that explains adding a command line handler, network variables (client and server-controlled), network transforms for auto-movement, and RPC introduction.
---

> 

Author: updated the intro section to make this a little clearer 

In this tutorial we will build on the work we have already done in Hello World and add in a few more features. As there are some minor differences between this tutorial and the Hello World series we will be starting afresh. You do not have to have completed either of the Hello World tutorials to complete this one, however, if you are new to Unity then we recommended that you  complete them in order to familiarise yourself with Unity. 

We will be covering the following:

- Adding a command line handler (to make launching easier)
- Network variables (client and server-controlled) 
- Network transforms
- RPCs

:::funfact

In the context of software or information modeling, a Golden Path (sometimes called happy path) is a default scenario featuring no exceptional or error conditions.
:::

In this guide we will build on the work we have already done in Hello World and add in a few more features, we will be covering the following:

- Adding a command line handler (to make launching easier)
- Network variables (client and server-controlled) 
- Network transforms
- RPCs

As there are some minor differences between this tutorial and the hello world one we will be starting afresh.

:::contribution Special Thanks

 This guide would not have been possible without the hard work and support of Matt Walsh. 

> :::

Author: updated and moved to bottom of page to avoid admonition clutter

## Create a new project in Unity
:::important
This tutorial requires an MLAPI-supported version of Unity (2019.4+).

> :::

Author: Updated to include a requirements section to all three tutorials and added some new intro text to Hello World tutorial as well.

A "Hello World" program is a computer program that outputs or displays the message "Hello, World!". Normally it is  the first program written by people learning to code. It  is also  used as a sanity test to make sure that a computer language is correctly installed, and that the operator understands how to use it.

This "Hello World" tutorial walks you through creating a project, installing the MLAPI package, and creating the basic components for your first networked game.

1. Open the Unity Hub.
1. Click New. 
1. Select type ‘3D’
1. Rename the project **GoldenPath**.
1. Select the location to save the project.

:::note
You will need to remember this path for later in the [guide](#creating-a-command-line-helper), when you are testing building the project from the command line. For this guide our path was `~/dev/mlapi-golden-path/`
:::

<iframe src="https://www.youtube.com/embed/AOZE-b9Q8R8?playlist=AOZE-b9Q8R8&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

## Import MLAPI Package via URL

> See the [Install MLAPI](../migration/installation.md) guide to install the MLAPI package.

Author: this is a tutorial that introduces different concepts from Hello World I have corrected the title and introduction to reflect the difference

## Creating Network Manager and selecting the Transport

1. Right click in the Hierarchy tab of the Main Unity Window.
1. Select **Create Empty**.
1. Rename the GameObject **NetworkManager**.
   
  :::tip

>   We renamed the GameObject because:

Author: Done

    * It makes it  easier to refer to later.
    * There should only be one **NetworkManager**, this is the object that contains the  `NetworkManager` component.
  :::

>   You have now created a new GameObject called **NetworkManager**.

Author: updated

## Creating an object to spawn for each connected player

1. Create 3D Object->Capsule (name it **Player**). 
1. Add a `NetworkObject` component.
1. Click the Assets folder.
1. Create a new Folder and call it **Prefabs**.
1. Make **Player** a prefab by dragging it to **Prefabs** folder you just created.
1. Delete **Player** from scene.

  :::tip
  We remove **Player**, because we will be using the network library to spawn the player. The library cannot track objects that start in the scene.
  :::

1. Select `NetworkManager`.
1. Inside the `NetworkManager` component tab, locate the  `NetworkPrefabs` field. 
1. Click `+` to create a slot
1. Drag this player prefab from above into the new empty slot

  :::tip
  When you select the **Default Player Prefab** , you are telling the library that when a client connect to the game, automatically spawn this prefab as the character for the connecting client. If you do not have the default selected for any prefab the game will crash on client connect.
  :::

  :::note
   You may see the following error reported `There is no NetworkPrefab Marked as a PlayerPrefab`. Once you have completed the above steps you can clear the error.
  :::

> 1. Create a 3D Object->Plane, centered at (0,0,0).

Reviewer: I'll bring this up again here since it was mentioned for the Hello World. This step isn't technically necessary. Removing this, however, involves cleaning up some lines below which reference a "Plane" so I'll leave it under your discretion.

> 1. Save your scene

Author: done

### Testing the basic network building blocks

1. To do a quick test you can click play.  
1. The Editor will start, and you will just see the plane.  
1. Without stopping the editor's play mode, navigate to the `NetworkManager` component in the Hierarchy tab (it will be underneath `DontDestroyOnLoad`).  
1. In the `NetworkManager` inspector scroll down and find the `Start Host` button.  
1. If you click it, you will see the player capsule spawn. 
1. Stop the player.

<iframe src="https://www.youtube.com/embed/iLb00icvRMs?playlist=iLb00icvRMs&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

## Creating a command line helper
This command line helper will help us launch our project from a command line.

1. Click the Assets folder.
1. Create a new Folder and call it **Scripts**.
1. Create a script called `NetworkCommandLine`.

> 1. Right click on the `NetworkManager` in the hierarchy view.
> 1. Create an empty GameObject underneath it.
> 1. Rename it `NetworkCommandLine`.

Reviewer: ```suggestion
1. Right click on the `NetworkManager` in the hierarchy view, and choose "Create Empty". This will create an empty GameObject, with `NetworkManager` as its parent.
1. Rename this child GameObject `NetworkCommandLine`.
```

Author: Done

> 1. In inspector  for `NetworkCommandLine` Click **Add Component** 

Reviewer: ```suggestion
1. In the inspector for `NetworkCommandLine`, select **Add Component**.
```

Author: Click a button

1. Click **Scripts** and add the `NetworkCommandLine.cs` script you created earlier.
1. Open the `NetworkCommandLine.cs` script.
1. Edit the `NetworkCommandLine.cs` script to match the following.

:::tip 
You can copy the script from here and paste it into your file.
   1. Select the code sample.
   1. Click **Copy** in the top right corner.

>    1. Paste it into your code editor.

Author: Done

       if (Application.isEditor) return;

       var args = GetCommandlineArgs();

       if (args.TryGetValue("-mlapi", out string mlapiValue))
       {
           switch (mlapiValue)
           {
               case "server":
                   netManager.StartServer();
                   break;
               case "host":
                   netManager.StartHost();
                   break;
               case "client":
         
                   netManager.StartClient();
                   break;
           }
       }
   }

   private Dictionary<string, string> GetCommandlineArgs()
   {
       Dictionary<string, string> argDictionary = new Dictionary<string, string>();

       var args = System.Environment.GetCommandLineArgs();

       for (int i = 0; i < args.Length; ++i)
       {
           var arg = args[i].ToLower();
           if (arg.StartsWith("-"))
           {
               var value = i < args.Length - 1 ? args[i + 1].ToLower() : null;
               value = (value?.StartsWith("-") ?? false) ? null : value;

               argDictionary.Add(arg, value);
           }
       }
       return argDictionary;
   }
}
```

</details>

> 1. Select  **File > Build Settings > Player Settings...**

Author: done

> 1. Change `Full Screen` mode into `Windowed` mode.  

Reviewer: ```suggestion
1. Under "Resolution and Presentation", change `Fullscreen Window` mode into `Windowed` mode.  
```

Author: Done

1. Save your scene.

   
:::tip
   If you are on a Pro Unity license, you may want to disable the splash screen
:::

<iframe src="https://www.youtube.com/embed/2swybHUigM8?playlist=2swybHUigM8&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

### Testing the command line helper
1. Select **File > Build and Run**. 
1. Create a folder called `Build`.

> 1. Name the binary `GoldenPath`.  

Reviewer: This is for Mac. On Windows it just wants the destination folder for the build. So for parity, Windows users should drop the build under the path "<path to project>Build/". Can't recall if this is different on Mac, but Unity on Windows uses the `Product Name` under Player settings as the executable name.

Author: have modified the text to remove the reference to two players and added the Windows instructions.

1. Your project will build, and it will launch, and you should see the plane.  
1. Quit your app.
1. Let's launch from the command line.  
   
#### For Windows

> 

Author: Content now provided, I am leaving it as server not host as i have removed the references to two players spawning and I do  not want to redo all the videos.

1. Open CMD
1. Enter the following:
    Server: `<path to project>/Build/GoldenPath.exe -mlapi server`
    Client: `<path to project>/Build/GoldenPath.exe -mlapi client`

#### For Mac

1. Open terminal

> 1. Enter the following `~/dev/mlapi-golden-path/GoldenPath/Build/GoldenPath.app/Contents/MacOS/GoldenPath -mlapi server -logfile - & ; ~/dev/mlapi-golden-path/GoldenPath/Build/GoldenPath.app/Contents/MacOS/GoldenPath -mlapi client -logfile -` 

Reviewer: Are the parameters ` -logfile - & ;` for server and `-logfile -` for the client necessary? They aren't being filtered inside of `NetworkCommandLine.cs`.

Author: I think they just stop Unity creating a logfile so it more a trimming the garbage for the command and so you dont end up with loads of logfiles on  your disks when testing or doing these tutorials

> 1. Enter the following `~/dev/mlapi-golden-path/GoldenPath/Build/GoldenPath.app/Contents/MacOS/GoldenPath -mlapi server -logfile - & ; ~/dev/mlapi-golden-path/GoldenPath/Build/GoldenPath.app/Contents/MacOS/GoldenPath -mlapi client -logfile -` 

Author: updated all these to be code blocks.

> You should see two players spawn.  Both should show a plane and a capsule (the capsule being the single player that was spawned)  

Reviewer: The instructions are to boot up a "Server" and a "Client". A server will not spawn a player. A host will. A client that connects to a server will have its player spawned on both clients.

Another catch here is that each build window will look identical. This is due to MLAPI spawning players at (0,0,0). So, I think reassuring the users that it is indeed working as intended would be helpful.

> You should see two players spawn.  Both should show a plane and a capsule (the capsule being the single player that was spawned)  

Author: I have updated this to better reflect what is actually happening. and left the code as it was as i do not want to redo all the videos.

Both should show a plane and a capsule (the capsule being the single player that was spawned).

<iframe src="https://www.youtube.com/embed/84wxdetXUQw?playlist=84wxdetXUQw&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

## Introducing a Server-controlled Network Variable 
 
1. Open the Scripts Folder.
1. Create a script called `NetworkVariableTest`.
1. Click the **Player** prefab.

> 1. In inspector for the **Player** prefab click **Add Component** 

Reviewer: Just to add the trailing period here too.

> 1. In inspector for the **Player** prefab click **Add Component** 

Author: Done

> 1. Click **Scripts** and add the `NetworkVariableTest.cs` script you created earlier

Author: Done

        {
            ServerNetworkVariable.Value = 0.0f;
            Debug.Log("Server's var initialized to: " + ServerNetworkVariable.Value);
        }
        else if (IsClient)
        {
            ClientNetworkVariable.Value = 0.0f;
            Debug.Log("Client's var initialized to: " + ClientNetworkVariable.Value);
        }
    }

    void Update()
    {
        var t_now = Time.time;
        if (IsServer)
        {
            ServerNetworkVariable.Value = ServerNetworkVariable.Value + 0.1f;
            if (t_now - last_t > 0.5f)
            {
                last_t = t_now;
                Debug.Log("Server set its var to: " + ServerNetworkVariable.Value + ", has client var at: "  + 
                    ClientNetworkVariable.Value);
            }
        }
        else if (IsClient)
        {
            ClientNetworkVariable.Value = ClientNetworkVariable.Value + 0.1f;
            if (t_now - last_t > 0.5f)
            {
                last_t = t_now;
                Debug.Log("Client set its var to: " + ClientNetworkVariable.Value + ", has server var at: "  + 
                    ServerNetworkVariable.Value);
            }
        }
    }
}
```
</details>

> 1. Save your scene.

Author: Done

<iframe src="https://www.youtube.com/embed/JsEVE6UaBIU?playlist=JsEVE6UaBIU&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

### Testing Server-controlled Network Variables 

1. Select **File > Build and Run**. 
1. Stop the player 

> 1. Launch the client & server together in a terminal as shown in [Testing the command line helper](#testing-the-command-line-helper). 

Author: Done

1. After a brief delay, the client and server will spawn.  

> 1. You should expect to see this in the console, showing that the server and client are sharing the variable. 

Author: Done

```
Server's var initialized to: 0
Client's var initialized to: 0
Server set its var to: 0.1, has client var at: 0
Client set its var to: 0.1, has server var at: 0.2
Server set its var to: 3.099999, has client var at: 2.6
Client set its var to: 3.099999, has server var at: 3.199999
Server set its var to: 6.099997, has client var at: 5.599997
```
:::note

> Since the printing doesn't happen on every tick, the numbers won't match up perfectly.

Author: messages do generate on every tick but the printing to the log doesn't so they dont match up, I will clarify in the text.

:::note
Since the printing to the terminal  does not happen on every tick, the numbers will not match up perfectly.
:::

:::

<iframe src="https://www.youtube.com/embed/cUvZ3m0idpM?playlist=cUvZ3m0idpM&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

## Introducing Network Transform

1. Click **Player** prefab. 
1. Click **Add Component** in the Inspector Tab.
1. Select **MLAPI** from the list shown.
1. Select `Network Transform` from the list shown.
1. Open the Scripts Folder.
1. Create a script called `NetworkTransformTest`.
1. Click the **Player** prefab.
1. In inspector for the **Player** prefab click **Add Component** 

> 1. Click **Scripts** and add the `NetworkTransformTest.cs` script you created earlier

Author: Done

1. Open the `NetworkTransformTest.cs` script.
1. Edit the `NetworkTransformTest.cs` script to match the following.

<details open>
<summary>Click to show/hide the Code.

</summary>

```csharp
using System;
using MLAPI;
using UnityEngine;

public class NetworkTransformTest : NetworkBehaviour
{
    void Update()
    {
        if (IsClient)
        {
            float theta = Time.frameCount / 10.0f;
            transform.position = new Vector3((float) Math.Cos(theta), 0.0f, (float) Math.Sin(theta));
        }
    }
}
```
</details>

> 1. Save your scene.

Author: Done

<iframe src="https://www.youtube.com/embed/TJkK3YTIkUs?playlist=TJkK3YTIkUs&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

### Testing Network Transform

1. Select **File > Build and Run**. 

> 1. Stop the player 

Reviewer 2: ```suggestion
1. Stop the player.
```

> 1. Launch the client & server together in a terminal as shown in [Testing the command line helper](#testing-the-command-line-helper). 

Reviewer 2: ```suggestion
1. Launch the client and server together in a terminal as shown in [Testing the command line helper](#testing-the-command-line-helper). 
```

Author: Done

1. After a brief delay, the client and server will spawn. 

> 1. You should see the player capsule moving in a circle on both the client and the server

Reviewer 2: ```suggestion
1. You should see the player capsule moving in a circle on both the client and the server.
```

Author: Done

<iframe src="https://www.youtube.com/embed/CQN_YUqdPfU?playlist=CQN_YUqdPfU&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

## Introducing RPCs

1. Open the Scripts Folder.
1. Create a script called `RpcTest`.
1. Click the **Player** prefab.

> 1. In inspector for the **Player** prefab click **Add Component** 

Reviewer 2: ```suggestion
1. In inspector for the **Player** prefab, click **Add Component**.
```

1. Click **Scripts** and add the `RpcTest.cs` script you created earlier
1. Right Click **Player** prefab. 
1. Open the `RpcTest.cs` script.
1. Edit the `RpcTest.cs` script to match the following.

<details open>
<summary>Click to show/hide the Code.

</summary>

```csharp
using MLAPI;
using MLAPI.Messaging;
using UnityEngine;

public class RpcTest : NetworkBehaviour
{
    private bool firstTime = true;

    [ClientRpc]
    void TestClientRpc(int value)
    {
        if (IsClient)
        {
            Debug.Log("Client Received the RPC #" + value);
            TestServerRpc(value + 1);
        }
    }

    [ServerRpc]
    void TestServerRpc(int value)
    {
        if (IsServer)
        {
            Debug.Log("Server Received the RPC #" + value);
            TestClientRpc(value);
        }
    }
    
    // Update is called once per frame
    void Update()
    {
        if (IsClient && firstTime)
        {
            firstTime = false;
            TestServerRpc(0);
        }
    }
}
```
</details>

> 1. Save your scene.

Reviewer 2: should be 9?
```suggestion
9. Save your scene.
```

<iframe src="https://www.youtube.com/embed/UEYj16CouLE?playlist=UEYj16CouLE&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

### Testing RPCs

1. Select **File > Build and Run**. 
1. Stop the player 
1. Launch the client & server together in a terminal as shown in [Testing the command line helper](#testing-the-command-line-helper). 
1. After a brief delay, the client and server will spawn.  
1. In the console, you should expect to see the client and server sending RPC messages to each other. 
1. The client kicks off the exchange in its `Update` call the first time with a counter value of 0.  

> 1. It then makes an RPC call to the server with the next value.  The server receives this and calls the client, etc.  In the console view, you will see:

Reviewer 2: ```suggestion
1. It then makes an RPC call to the server with the next value.  The server receives this and calls the client. In the console view, you will see:
```

Author: Done

```
Server Received the RPC #1
Client Received the RPC #1
Server Received the RPC #2
Client Received the RPC #2
Server Received the RPC #3
Client Received the RPC #3
...
```
<iframe src="https://www.youtube.com/embed/0mSciPQauyw?playlist=0mSciPQauyw&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

:::note Congrats!

> Congratulations you have learnt how to use some of the basic building blocks of Multiplayer game creation. For more information on the relevant concepts intorduced here please refer to the following sections of the documentaiton:

Reviewer 2: ```suggestion
Congratulations you have learned how to use some of the basic building blocks of Multiplayer game creation. For more information on the relevant concepts introduced here please refer to the following sections of the documentation:
```

Author: Done

- [Network variables (client and server-controlled)](../mlapi-basics/networkvariable.md)
- [Network transforms](../components/networktransform.md)

> - [RPCs](../advanced-topics/message-system/about-rpc.md)

Reviewer 2: may want to make this a see also or next steps section with a heading?

Author: Done

---
id: goldenpath
title:  Building Golden Path
description: Tutorial that explains adding a command line handler, network variables (client and server-controlled), network transforms for auto-movement, and RPC introduction.
---

In this tutorial we will build on the work we have already done in Hello World and add in a few more features. As there are some minor differences between this tutorial and the Hello World series we will be starting afresh. You do not have to have completed either of the Hello World tutorials to complete this one, however, if you are new to Unity then we recommended that you  complete them in order to familiarise yourself with Unity. 

We will be covering the following:

- Adding a command line handler (to make launching easier)
- Network variables (client and server-controlled) 
- Network transforms
- RPCs

:::funfact

In the context of software or information modelling, a Golden Path (sometimes called happy path) is a default scenario featuring no exceptional or error conditions.
:::

## Requirements

This tutorial requires an MLAPI-supported version of Unity (2019.4+).

## Create a new project in Unity

1. Open the Unity Hub.

> 1. Click New. 
> 1. Select type ‘3D’

Reviewer 2: ```suggestion
1. Click **New**. 
1. Select type ‘3D’.
```

Author: Done

1. Rename the project **GoldenPath**.
1. Select the location to save the project.

:::note

> Make note of this path. You will need it  later in the [tutorial](#creating-a-command-line-helper), when you are testing building the project from the command line. For this guide we used a MAC and our path was `~/dev/mlapi-golden-path/` on a MAC your path may be slightly different.

Reviewer 2: Do you mean it may differ on a Windows machine or just on their machine? Also MAC should be Mac. MacOS or OSX are names for the operating system.
```suggestion
Make note of this path. You will need it  later in the [tutorial](#creating-a-command-line-helper), when you are testing building the project from the command line. For this guide, we used a Mac and our path was `~/dev/mlapi-golden-path/`. On a Windows machine, your path may be slightly different.
```

Author: Done

1. Select **NetworkManager**.
1. Inside the **NetworkManager** Component tab, locate the  `NetworkPrefabs` field. 
1. Click `+` to create a slot
1. Drag this player prefab from above into the new empty slot

  :::important
  When you select the **Default Player Prefab** , you are telling the library that when a client connects to the game, automatically spawn this prefab as the character for the connecting client. If you do not have the default selected for any prefab the game will crash on client connect.
  :::

  :::note
   You may see the following error reported `There is no NetworkPrefab Marked as a PlayerPrefab`. Once you have completed the above steps you can clear the error.
  :::

1. Create a **3D Object->Plane**, centered at (0,0,0).
1. Save your scene.

<iframe src="https://www.youtube.com/embed/B_FWb4J1Pxw?playlist=B_FWb4J1Pxw&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

### Testing the basic network building blocks

> 1. To do a quick test you can click play.  

Reviewer 2: Definitely may want to include a stem for future guidelines. Also would not use quick, cause if it's not quick....

```suggestion
To perform a test:

1. Click **Play**.  
```

Author: Done

1. The Editor will start, and you will just see the plane.  
1. Without stopping the editor's play mode, navigate to the `NetworkManager` component in the Hierarchy tab (it will be underneath `DontDestroyOnLoad`).  
1. Inside the `NetworkManager` Inspector tab. scroll down and find the `Start Host` button.  
1. If you click it, you will see the player capsule spawn. 
1. Stop the player.

<iframe src="https://www.youtube.com/embed/iLb00icvRMs?playlist=iLb00icvRMs&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

## Creating a command line helper

> This command line helper will help us launch our project from a command line.

Reviewer 2: ```suggestion
This command line helper will launch our project from a command line.
```

       var args = System.Environment.GetCommandLineArgs();

       for (int i = 0; i < args.Length; ++i)
       {
           var arg = args[i].ToLower();
           if (arg.StartsWith("-"))
           {
               var value = i < args.Length - 1 ? args[i + 1].ToLower() : null;
               value = (value?.StartsWith("-") ?? false) ? null : value;

               argDictionary.Add(arg, value);
           }
       }
       return argDictionary;
   }
}
```

</details>

10. Select  **File > Build Settings > Player Settings...**
1. In the **Resolution and Presentation** tab change `Fullscreen Window` mode into `Windowed` mode.    
1. Save your scene.

   
:::tip
   If you are on a Pro Unity license, you may want to disable the splash screen
:::

<iframe src="https://www.youtube.com/embed/2swybHUigM8?playlist=2swybHUigM8&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

> ### Testing the command line helper

Reviewer 2: With guidelines coming, may want to add intro lines? Otherwise will have to later.

```suggestion
### Testing the command line helper

To test the command line:
```

Author: intro lines added to all three tutorials

1. Select **File > Build and Run**. 
1. Create a folder called `Build`.
1. Name the binary `GoldenPath`.  
1. Your project will build, and it will launch, and you should see the plane.  
1. Quit your app.

> 1. Let's launch from the command line.  

Reviewer 2: do you mean they should enter a command here? or the next step for this is the windows or mac instructions below?

```suggestion
1. Launch from the command line on your Windows or Mac machine (see below).
```

Author: Done

   
#### For Windows

> 

Reviewer 2: Same again on intro lines:

```suggestion
To test the command line on Windows:
```

Author: Done

1. Open CMD
1. Enter the following:
   
    Server: 
    ```
    <path to project>/Build/GoldenPath.exe -mlapi server
    ```
    Client: 
    ```
    <path to project>/Build/GoldenPath.exe -mlapi client
    ```

#### For Mac

> 
> 1. Open Terminal
> 1. Enter the following

Reviewer 2: ```suggestion
To test the command line on Mac:

1. Open Terminal.
1. Enter the following:
```

Author: Intro line covers this

   ```
   ~/dev/mlapi-golden-path/GoldenPath/Build/GoldenPath.app/Contents/MacOS/GoldenPath -mlapi server -logfile - & ; ~/dev/mlapi-golden-path/GoldenPath/Build/GoldenPath.app/Contents/MacOS/GoldenPath -mlapi client -logfile -
   ```

Both should show a plane and a capsule (the capsule being the single player that was spawned). 

<iframe src="https://www.youtube.com/embed/84wxdetXUQw?playlist=84wxdetXUQw&loop=1&&autoplay=0&controls=1&showinfo=0&mute=1"   width="854px"
        height="480px" className="video-container" frameborder="0" position="relative" allow="accelerometer; autoplay; loop; playlist; clipboard-write; encrypted-media; gyroscope; picture-in-picture"  allowfullscreen=""></iframe>

> 
> 
> 
> ## Introducing a Server-controlled Network Variable 
>  
> 1. Open the **Scripts** Folder.
> 1. Create a script called `NetworkVariableTest`.
> 1. Click the **Player** prefab.
> 1. In the **Player** prefab Inspector tab., click **Add Component**. 
> 1. Click **Scripts**, and add the `NetworkVariableTest.cs` script you created earlier.
> 1. Open the `NetworkVariableTest.cs` script.
> 1. Edit the `NetworkVariableTest.cs` script to match the following.
> 

Reviewer 2: need intro and tweaks
```suggestion
## Introducing a Server-controlled Network Variable 

Create and edit a test script to introduce a server-controlled `NetworkVariable`:
 
1. Open the *Scripts* folder.
1. Create a script called `NetworkVariableTest`.
1. Click the **Player** prefab.
1. In the **Player** prefab Inspector tab., click **Add Component**. 
1. Click **Scripts**, and add the `NetworkVariableTest.cs` script you created earlier.
1. Open the `NetworkVariableTest.cs` script.
1. Edit the `NetworkVariableTest.cs` script to match the following.

```

Author: Section updated
